src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,128 @@
+ * size of the vector is known at compile time, a `TempBuffer` object should be
+ * used instead.
+ *
+ * If needed it should be fairly straightforward to generalize to
+ * `ComplexDataVector`.
+ */
+
+template <typename T>
+class DynamicBuffer {
+ public:
+  static constexpr bool is_data_vector_type =
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+
+  DynamicBuffer() = default;
+
+  /*!
+   * Constructs a `DynamicBuffer`. The `number_of_vectors` corresponds to the
+   * number of `DataVector`s which are saved inside, each of which has size
+   * `number_of_grid_points`. `number_of_grid_points` has to be 1 if T is a
+   * fundamental type.
+   */
+  DynamicBuffer(const size_t number_of_vectors,
+                const size_t number_of_grid_points) noexcept
+      : number_of_grid_points_(number_of_grid_points),
+        data_vectors_(number_of_vectors) {
+    if constexpr (is_data_vector_type) {
+      buffer_.resize(number_of_vectors * number_of_grid_points);
+      set_references();
+    } else {
+      static_assert(
+          std::is_fundamental_v<T>,
+          "T was found to be neither a DataVector nor a fundamental type. "
+          "`DynamicBuffer` is not implemented yet for complex numbers.");
+      if (number_of_grid_points != 1) {
+        ERROR(
+            "DynamicBuffer must have number_of_grid_points == 1 when T is a "
+            "fundamental type but has number_of_grid_points = "
+            << number_of_grid_points_);
+      }
+    }
+  }
+
+  ~DynamicBuffer() = default;
+  DynamicBuffer(DynamicBuffer&& other) noexcept = default;
+
+  DynamicBuffer(const DynamicBuffer& other) noexcept
+      : number_of_grid_points_(other.number_of_grid_points_) {
+    if constexpr (is_data_vector_type) {
+      data_vectors_.resize(other.size());
+      buffer_ = other.buffer_;
+      set_references();
+    } else {
+      data_vectors_ = other.data_vectors_;
+    }
+  }
+
+  // copy-and-swap: acts as both move and copy assignment operator
+  DynamicBuffer& operator=(DynamicBuffer other) noexcept {
+    std::swap(*this, other);
+    return *this;
+  }
+
+  T& operator[](size_t index) { return data_vectors_[index]; }
+  T& at(size_t index) { return data_vectors_.at(index); }
+  const T& operator[](size_t index) const { return data_vectors_[index]; }
+  const T& at(size_t index) const { return data_vectors_.at(index); }
+
+  auto begin() { return data_vectors_.begin(); }
+  auto end() { return data_vectors_.end(); }
+  auto begin() const { return data_vectors_.begin(); }
+  auto end() const { return data_vectors_.end(); }
+
+  size_t size() const noexcept { return data_vectors_.size(); }
+
+  /// \cond
+  // clang-tidy: runtime-references
+  void pup(PUP::er& p) noexcept {
+    p | number_of_grid_points_;
+    p | buffer_;
+    if constexpr (is_data_vector_type) {
+      if (p.isUnpacking()) {
+        set_references();
+      }
+    } else {
+      p | data_vectors_;
```

Comment:
does this actually do anything? I thought `data_vectors_` should be empty in this case so why bother sending it?

Reply:
in case of fundamental types, all the data is saved in `data_vectors_` and the buffer is empty, so that all the iterators and similar still work.

Reply:
right. I don't have a great name for `data_vectors_`, maybe just `data_` to make it more accurate when `T` is not a `DataVector`? :shrug: up to you